examples/step-81/step-81.cc

```diff
@@ -0,0 +1,894 @@
+                                            update_quadrature_points |
+                                            update_normal_vectors |
+                                            update_JxW_values);
+
+    const unsigned int dofs_per_cell = fe->dofs_per_cell;
+
+    const unsigned int n_q_points      = quadrature_formula.size();
+    const unsigned int n_face_q_points = face_quadrature_formula.size();
+
+    FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+    Vector<double>     cell_rhs(dofs_per_cell);
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    // This is assembling the interior of the domain on the left hand side.
+    // So we are assembling
+    // \f{align*}{
+    // \int_\Omega (\mu_r^{-1}\nabla \times \varphi_i) \cdot
+    // (\nabla\times\bar{\varphi}_j)\text{d}x
+    // - \int_\Omega \varepsilon_r\varphi_i \cdot \bar{\varphi}_j\text{d}x
+    // \f}
+    // and
+    // \f{align}{
+    // i\int_\Omega J_a \cdot \bar{\varphi_i}\text{d}x
+    // - \int_\Omega \mu_r^{-1} \cdot (\nabla \times \bar{\varphi_i}) \text{d}x.
+    // \f}
+    // In doing so, we need test functions $\varphi_i$ and $\varphi_j$, and the
+    // curl of these test variables. We must be careful with the signs of the
+    // imaginary parts of these complex test variables. Moreover, we have a
+    // conditional that changes the parameters if the cell is in the PML region.
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        fe_values.reinit(cell);
+        FEValuesViews::Vector<dim> real_part(fe_values, 0);
+        FEValuesViews::Vector<dim> imag_part(fe_values, dim);
+
+        cell_matrix = 0.;
+        cell_rhs    = 0.;
+
+        cell->get_dof_indices(local_dof_indices);
+        const auto id = cell->material_id();
+
+        const auto &quadrature_points = fe_values.get_quadrature_points();
+
+        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+          {
+            const Point<dim> &position = quadrature_points[q_point];
+            const auto        radius   = position.norm();
+            const auto inner_radius    = perfectly_matched_layer.inner_radius;
+
+            auto       mu_inv  = parameters.mu_inv(position, id);
+            auto       epsilon = parameters.epsilon(position, id);
+            const auto J_a     = parameters.J_a(position, id);
+
+            if (radius >= inner_radius)
+              {
+                auto A = perfectly_matched_layer.a_matrix(position);
+                auto B = perfectly_matched_layer.b_matrix(position);
+                auto d = perfectly_matched_layer.d(position);
+
+                mu_inv  = mu_inv / d;
+                epsilon = invert(A) * epsilon * invert(B);
+              };
+
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+              {
+                const auto phi_i = real_part.value(i, q_point) -
+                                   1.0i * imag_part.value(i, q_point);
```

Comment:
Could you explain why we have here a negative sign and in `phi_j` a positive one? Maybe it is a stupid question^^